### PR TITLE
Overvote bug

### PR DESCRIPTION
**Pull request by jonahkagan**
Task: #607 

- Fixes a bug where deselecting a contest choice didn't reset the
interpretation to null, so you could accidentally send a `VOTE`
interpretation with no `choiceIds`.
- Adds some tests to start to cover the ballot auditing flow more
thoroughly
- Adds some new test helpers for rendering and mocking the api
(explained in more detail in the comments)

### CODE REVIEW

File: client/src/components/testUtilities.tsx [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  )
+
+  return {
+    ...render(ui, { wrapper: Wrapper }),
+    // Adding `history` to the returned utilities to allow us
+    // to reference it in our tests (just try to avoid using
+    // this to test implementation details).
+    history,
+  }
+}
```
Comment (jonahkagan) [inline client/src/components/testUtilities.tsx:64]: @MorganLove here's one new test helper I found in the docs. Seems like a good starting point for simplifying rendering components that rely on route params.

File: client/src/components/testUtilities.tsx [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+      options,
+    }))
+    const expectedCalls = apiCalls.map(({ endpoint, options }) => ({
+      endpoint,
+      options,
+    }))
+    expect(actualCalls).toEqual(expectedCalls)
+    mock.mockReset()
+  }
+}
```
Comment (jonahkagan) [inline client/src/components/testUtilities.tsx:94]: @MorganLove here's a new idea for mocking api calls

File: client/src/components/DataEntry/BallotAudit.tsx [hunk 3]
```diff
@@ -120,17 +120,14 @@ const BallotAuditContest = ({
...
       } else {
+        const choiceIds = checked
+          ? [...interpretation.choiceIds, value]
+          : interpretation.choiceIds.filter(v => v !== value)
         setInterpretation({
           ...interpretation,
-          interpretation: Interpretation.VOTE,
-          choiceIds: interpretation.choiceIds.filter(v => v !== value),
+          interpretation: choiceIds.length > 0 ? Interpretation.VOTE : null,
+          choiceIds,
```
Comment (jonahkagan) [inline client/src/components/DataEntry/BallotAudit.tsx:130]: Here's the actual bug fix

File: client/src/components/DataEntry/index.test.tsx [hunk 4]
```diff
@@ -265,5 +257,114 @@ describe('DataEntry', () => {
...
+            {
+              contestId: 'contest-id-2',
+              interpretation: 'CANT_AGREE',
+              choiceIds: [],
+              comment: null,
+            },
+          ],
+        }),
+        apiCalls.ballotsOneAudited,
+      ])
```
Comment (jonahkagan) [inline client/src/components/DataEntry/index.test.tsx:284]: Here's how that new mock helper gets used

File: client/src/components/DataEntry/index.test.tsx [hunk 5]
```diff
@@ -265,5 +257,114 @@ describe('DataEntry', () => {
...
+      expect(screen.queryByText('Choice Three')).not.toBeInTheDocument()
+      expect(screen.queryByText('Choice Four')).not.toBeInTheDocument()
+
+      // Submit the ballot
+      userEvent.click(
+        screen.getByRole('button', { name: 'Submit & Next Ballot' })
+      )
+      await screen.findByText('Auditing ballot 3 of 27')
+
+      checkMockApi()
```
Comment (jonahkagan) [inline client/src/components/DataEntry/index.test.tsx:325]: and then you call this function to run the assertions at the end

File: client/src/components/DataEntry/index.test.tsx (hunk 4 above)
Comment (MorganLove) [reply client/src/components/DataEntry/index.test.tsx:284]: Interesting! It looks more neat, but it's also not immediately apparent what's going on. I'll have to dig into it and play with it myself before I grok it completely.

**Comment by MorganLove** [comment]
@jonahkagan Lots of changes! I like the test innovations. The test failure right now looks like it's some significant snapshot changes, so probably should look through them to make sure they are desired and not something broken.

**Comment by jonahkagan** [comment]
@MorganLove ok I cleaned up this PR. Also changed the API mocking up a bit and added a comment to try to make it clearer. Let me know your thoughts

**Comment by MorganLove** [comment]
@jonahkagan Okay, makes more sense now, thanks! So is the api mocking something that can be used to obsolete code like this: https://github.com/votingworks/arlo/blob/bb03d24b2c5f44e1a1d2adc60b38c060376ba24d/client/src/components/DataEntry/index.test.tsx#L114 ?

Because if we can use it to replace all those separate `mockImplementation`s that'd save sooo much time and effort I think.

**Comment by jonahkagan** [comment]
Yes I think so! I should go back and update the rest of the tests in this file. Just wanted to make sure you were into it before I did that work. I'll be curious to hear your thoughts more as we try this new strategy - I'm sure we'll find out its strengths and weaknesses and improve upon it.

**Comment by jonahkagan** [comment]
@MorganLove I went back and updated the tests to use the new API mocks. There were some race conditions happening with `jest.spyOn(utilities, 'api')` - the mocks were receiving calls from other tests that were running in parallel. I was able to solve it by mocking `window.fetch` instead, which I think is nice anyway because it's closer to the actual separation between the frontend and backend.

Can you give it another look before I merge?

**Comment by MorganLove** [comment]
@jonahkagan Looking good! There's still a couple tests failing on CI, but I much prefer this new method of managing the api mocking. :slightly_smiling_face:

**Pull request merged**